primitives/bridge/src/lib.rs

```diff
@@ -0,0 +1,385 @@
+pub const FROM_DARWINIA_LATEST_CONFIRMED_NONCE_METHOD: &str =
+	"FromDarwiniaInboundLaneApi_latest_confirmed_nonce";
+/// Name of the `FromDarwiniaInboundLaneApi::unrewarded_relayers_state` runtime method.
+pub const FROM_DARWINIA_UNREWARDED_RELAYERS_STATE: &str =
+	"FromDarwiniaInboundLaneApi_unrewarded_relayers_state";
+
+/// Name of the `ToDarwiniaOutboundLaneApi::estimate_message_delivery_and_dispatch_fee` runtime method.
+pub const TO_DARWINIA_ESTIMATE_MESSAGE_FEE_METHOD: &str =
+	"ToDarwiniaOutboundLaneApi_estimate_message_delivery_and_dispatch_fee";
+/// Name of the `ToCrabOutboundLaneApi::message_details` runtime method.
+pub const TO_DARWINIA_MESSAGE_DETAILS_METHOD: &str = "ToDarwiniaOutboundLaneApi_message_details";
+/// Name of the `ToDarwiniaOutboundLaneApi::latest_generated_nonce` runtime method.
+pub const TO_DARWINIA_LATEST_GENERATED_NONCE_METHOD: &str =
+	"ToDarwiniaOutboundLaneApi_latest_generated_nonce";
+/// Name of the `ToDarwiniaOutboundLaneApi::latest_received_nonce` runtime method.
+pub const TO_DARWINIA_LATEST_RECEIVED_NONCE_METHOD: &str =
+	"ToDarwiniaOutboundLaneApi_latest_received_nonce";
+
+/// Name of the `DarwiniaFinalityApi::best_finalized` runtime method.
+pub const BEST_FINALIZED_DARWINIA_HEADER_METHOD: &str = "DarwiniaFinalityApi_best_finalized";
+// === end
+
+/// Convert a 256-bit hash into an AccountId.
+pub struct AccountIdConverter;
+impl Convert<H256, AccountId> for AccountIdConverter {
+	fn convert(hash: H256) -> AccountId {
+		hash.to_fixed_bytes().into()
+	}
+}
+
+/// Darwinia chain.
+#[derive(RuntimeDebug)]
+pub struct Darwinia;
+impl Chain for Darwinia {
+	type BlockNumber = BlockNumber;
+	type Hash = Hash;
+	type Hasher = Hashing;
+	type Header = Header;
+	type AccountId = AccountId;
+	type Balance = Balance;
+	type Index = Nonce;
+	type Signature = Signature;
+}
+
+/// Crab chain.
+#[derive(RuntimeDebug)]
+pub struct Crab;
+impl Chain for Crab {
+	type BlockNumber = BlockNumber;
+	type Hash = Hash;
+	type Hasher = Hashing;
+	type Header = Header;
+	type AccountId = AccountId;
+	type Balance = Balance;
+	type Index = Nonce;
+	type Signature = Signature;
+}
+
+/// Message verifier that is doing all basic checks.
+///
+/// This verifier assumes following:
+///
+/// - all message lanes are equivalent, so all checks are the same;
+/// - messages are being dispatched using `pallet-bridge-dispatch` pallet on the target chain.
+///
+/// Following checks are made:
+///
+/// - message is rejected if its lane is currently blocked;
+/// - message is rejected if there are too many pending (undelivered) messages at the outbound
+///   lane;
+/// - check that the sender has rights to dispatch the call on target chain using provided
+///   dispatch origin;
+/// - check that the sender has paid enough funds for both message delivery and dispatch.
+#[derive(RuntimeDebug)]
+pub struct DarwiniaFromThisChainMessageVerifier<B, R>(PhantomData<(B, R)>);
```

Comment:
We should also expect an counter-party `CrabFromThisChainMessageVerifier` right? But I can not find it.